### ISSUE TITLE
QtWebEngine transition (again, but with a plan)

### ISSUE THREAD

**Issue by danieleds** [notepadqq/notepadqq#572]
@JuBan1 
@Teklad 

I've thought about the problem and I finally came up with this plan in order to solve it.

## Step 1: Asynchronous transition

In step 1 we'll slowly move our API from the current blocking design to an asynchronous one. This will involve all the "sendMessageWithResult" methods. The backend engine will stay the same, our loved QtWebKit.

## Step 2: Transition to QtWebEngine

Everything should be asynchronous now. It's just a matter of switching the engine, which is anyway well confined within the Editor class. Should be a smooth step.

## Step 3: Performance optimizations

This is were we'll try to take advantages of all the new features that QtWebEngine provides, in order to fix or improve the performance

-----------

**Why not our own custom native widget?** I thought a lot about it, but basically the reason is: we don't have the resources to build it, but, more importantly, to maintain it. We won't compete with the other editors *at all*, and it's like going back to the beginning. A web view based editor is the best we can get right now: big community behind, lot of languages, lot of themes, lot of features.

**How to make everything async in step 1?** I've prepared some code. It's not pretty from the "syntactic" point of view, but the implementation is (I think) quite simple and easy. I'll push a commit soon, as soon as I manage to clean the code a little bit.

I'd like to hear your opinion guys :)

**Comment by JuBan1** [comment]
I vaguely remember we've had a lot of trouble with the previous QWebEngine attempt, but perhaps this way it'll be a little more resistant to issues.

On the note of "everything async"- it might be a good idea to also have blocking variants since sometimes we might want to block execution until a function returns. Either non-async variants of functions or a `std::future`-like object that allows us to `.wait()` on the result of an execution.

Though I'll still pursue a native widget implementation on the side. I've been using a Nqq branch with my experimental widget for the past couple of weeks and -aside from only having four syntax-highlighted languages at the moment- I don't even notice that it's not CodeMirror anymore. A few key features are still missing but I think it's got promise. 

When I've got it in a presentable state (which will take me a while since I don't put a lot of work into it at the moment) I'd still like to present it for consideration. At worst it'll amount to some wasted time and valuable experiences.

**Comment by danieleds** [comment]
> When I've got it in a presentable state (which will take me a while since I don't put a lot of work into it at the moment) I'd still like to present it for consideration. At worst it'll amount to some wasted time and valuable experiences.

In case it grows, in the future we can switch to that. Or even try to provide two different backends for a while. For now, at least we'll be able to build an easy transition path away from qtwebkit asap.  
Can you show some screenshot?

**Comment by JuBan1** [comment]
> For now, at least we'll be able to build an easy transition path away from qtwebkit asap.

Yep, that's what I thought too.

> Can you show some screenshot?

Not that much to show. This is CodeMirror vs OTE next to each other:

![img](https://i.imgur.com/FptR5HA.png)

The big outstanding issues with it are:

* Multiple cursor selection (which I'm working on)

* Syntax highlighting (for which I'll probably implement a parser for the Highlight Definition format used by KDE's Kate since they have a [wealth of available definitions](https://cgit.kde.org/syntax-highlighting.git/tree/data/syntax) ).

* Code Folding (which I haven't touched yet)

**Comment by Teklad** [comment]
Once the native widget gets far enough along we could provide it as a build
option or just offer up a notepadqq fork or branch depending on how
@danieleds wants to go about that.  If it ever gets to the point that its
on par with the webengine counterpart we can always replace what's in the
main branch....  that is of course after extensive bug testing.  Once the
native widget is rather feature complete it will probably be easier to
maintain in the long run vs. praying for QWebEngine rendering to perform
equally across different versions of QT.  This would also allow supporting
older versions of QT.  QWebEngine being in its infancy still makes any kind
of real backwards compatibility with QT a moot point.

**Comment by JuBan1** [comment]
Compatibility is a bit of an issue perhaps. 

Ubuntu 16.04 LTS contains Qt 5.6 in its repo, and I believe this would probably be a good minimum to support.

However, a lot of functionality in QWebEngineView/Page has been introduced over time, with some things such as printing only being available in 5.7+.

**Comment by danieleds** [comment]
It's here if you want to give it a try:

https://github.com/notepadqq/notepadqq/commits/webengine2

The transition has gone really smooth. I didn't have to change basically anything outside of editor.cpp.

Currently the wrong context-menu is showing up, but everything else seems to be working. Travis fails because it runs on Ubuntu Trusty which doesn't have a package for QWebEngine.

**Comment by JuBan1** [comment]
I closed a number of our old QWebEngine issues that aren't applicable anymore. However, a few are still around with the webengine2 branch:

#405 #378 still exist for me.

**Comment by danieleds** [comment]
Yup, for me too. I want to investigate again #378 but if it's like @Teklad said, we can close it and/or find out what's the problematic language.

**Comment by danieleds** [comment]
In addition, on macOS the webengine2 branch works 100 times better than master. But we still have a few crashes in particular cases.

**Comment by danieleds** [comment]
About #405: https://bugreports.qt.io/browse/QTBUG-58324

Looks like another upstream issue. We can at most provide a dirty workaround.

**Comment by danieleds** [comment]
Travis has been fixed

**Comment by samikrc** [comment]
Any update on https://github.com/notepadqq/notepadqq/issues/461? Should I check out any of the recent builds? Thanks.

**Comment by danieleds** [comment]
@samikrc , the porting is not complete yet. You should try updating Qt to a more recent version in the meantime. Let us know if the problem disappears.

**Comment by danieleds** [comment]
### WebEngine status update

The branch "webengine2" is currently working and the code changes with respect to master are really minimal and clean. However it exhibits some crashes in peculiar situations, as well as some other issue with events. I highly suspect that this happens because we're doing nasty things with the event loop (entering and exiting it in a highly deep way).

For these reasons I'm currently updating the "webengine2-promise" branch too, in which the event loop doesn't get blocked anymore and we use a Promise/A+ -like interface in order to handle concurrency. This transition, however, requires more time because we have to update each call to the editor methods to the new API. The good news is, the event loop strategy and the promise strategy can live together at the same time, and in fact they do in the "webengine2-promise" branch. For this reason, we might be able to get it working very soon, actually as soon as the quirks go away. So we don't have to convert everything to Promises, at least for now.

**Comment by Teklad** [comment]
Just to touch up on this....  QWebEngine by design is **very** asynchronous.  It would probably be best to modify notepadqq to adhere to that design principle completely in the long term.  Using QEventLoop to block causes some really unusual side-effects in QT signals across threads.  This is compounded by the fact we use lambdas in slots for event handling.

Unfortunately, nqq was created way before QWebEngine was even a thing so we're looking at quite a few modifications to make it completely compliant.

To my understanding the "proper" way to communicate with QWebEngine is via slots.  Considering how nqq is setup now this is going to be a lot of work.  Not impossible, just very tedious.

I checked my old QWebEngine branch and it appears to have been completely broken by an update to QT at some point, so most of the code there which might have been useful for referencing is moot.  I did have some of that slot magic setup, but also had a lot of extra stuff which made the branch rather unusable.

Finding a way to make nqq not rely so much on waiting for a reply would have a lot of performance benefits.  I'm still debating on how to do it myself without turning the code base into a train wreck.

**Comment by JuBan1** [comment]
> converting each API is a boring work.

I believe in you!

**Comment by danieleds** [comment]
Are you guys working on any important fix that is going to need to be released quickly? I'm thinking about moving webengine2-promise to master very soon.

P.S. we'll be using c++14

**Comment by JuBan1** [comment]
Good with me.

**Comment by Teklad** [comment]
I approve

**Comment by danieleds** [comment]
Merging.

I'm going to do some terrible things to the git history... rebasing was impossible :(
Sorry git.

**Comment by danieleds** [comment]
Okay. Everything seems to be good for now.

I "officially" dropped support for Qt < 5.6, but I'm not sure what the best threshold is.

In addition, with this change we can forget about a lot of old distros like Ubuntu 14.04 and, as we can now use c++14, we're free to make the code a little better.

It appears that, in the end, we finally did it.

**Comment by JuBan1** [comment]
Is 5.6 necessary? Our stats show 5.5 is by far the most popular, because because it's in some Ubuntu/Debian LTS or so.

**Comment by Teklad** [comment]
So long as it works I don't have a problem with it.  Worst case scenario if we did use any included 5.6 features a simple #ifdef could easily enable/disable features based on QT version.... or do what we did with printing and provide a work-around.

**Comment by danieleds** [comment]
@JuBan1 5.5 is also popular because it is the version shipped with snap (which we decide)

In fact, if you hide it from the plot, Qt 5.6 is still far less used than 5.9. So, of we need to, it wouldn't be impossible to drop support for 5.6.

**Comment by danieleds** [comment]
No problem with lowering it to 5.5 though... As long as QtWebEngine is not too buggy
